apis/python/src/tiledbsoma/_dense_nd_array.py

```diff
@@ -71,6 +72,18 @@ class DenseNDArray(NDArray, somacore.DenseNDArray):
 
     __slots__ = ()
 
+    def non_empty_domain(self) -> Tuple[Tuple[int, int], ...]:
+        """
+        Retrieves the non-empty domain for each dimension, namely the smallest
+        and largest indices in each dimension for which the sparse array has
+        data occupied.  This is nominally the same as the domain used at
+        creation time, but if for example only a corner of the available domain
+        has actually had data written, this function will return a tighter
+        range.
+        """
+        with tiledb.open(self.uri, ctx=self.context.tiledb_ctx) as A:
+            return A.nonempty_domain()  # type: ignore
+
```

Comment:
What I recommend instead is to remove this, and add a `TileDBArray._tiledb_nonempty_domain()` function into [_tiledb_array.py](https://github.com/single-cell-data/TileDB-SOMA/blob/main/apis/python/src/tiledbsoma/_tiledb_array.py#L45). This uses the already opened handle and does not introduce a new `DenseNDArray.non_empty_domain`.

On this line, call `ned = self._handle._tiledb_nonempty_domain()` which then just keep getting the NED internal.
https://github.com/single-cell-data/TileDB-SOMA/pull/1810/files#diff-1c8b39ff256e708f483b5e9ebdb3c0dd07f14a78cd487cbf1e8d7c49373d9997R137

Reply:
Also minor nit, but I think we favor naming as `nonempty_domain` over `non_empty_domain`.

Reply:
@nguyenv it's already released for sparse (i'm adding it for dense) and i don't want a backward-incompatible name change

thanks though!

Reply:
OK nevermind about the name change. But I think the fact that both `SparseNDArray` and `DenseNDArray` use this is even more reason to add this into `TileDBArray` and then in both classes use

```
    def non_empty_domain(self) -> Tuple[Tuple[int, int], ...]:
        """
        Retrieves the non-empty domain for each dimension, namely the smallest and
        largest indices in each dimension for which the sparse array has data occupied.
        This is nominally the same as ``used_shape``, but if for example the
        leading/trailing rows/columns of the sparse array are entirely unoccupied, this
        function will return a tighter range.
        """
        return self._handle._tiledb_nonempty_domain()
```

Reply:
just pushed a commit w/ that change! :)

Reply:
```
>>> exp.obs._handle. <---- tab-complete
exp.obs._handle.close(       exp.obs._handle.metadata     exp.obs._handle.reader       exp.obs._handle.uri
exp.obs._handle.closed       exp.obs._handle.mode         exp.obs._handle.schema       exp.obs._handle.writer
exp.obs._handle.context      exp.obs._handle.open(        exp.obs._handle.timestamp_ms

>>> type(exp.obs._handle)
<class 'tiledbsoma._tdb_handles.ArrayWrapper'>

>>> exp.obs._handle._
exp.obs._handle._abc_impl          exp.obs._handle._flush_hack(       exp.obs._handle._opener(
exp.obs._handle._check_open(       exp.obs._handle._handle
exp.obs._handle._do_initial_reads( exp.obs._handle._is_protocol
```

so we would need a follow-on once we've gotten `nonempty_domain` wired in from C++